clang/test/Driver/riscv-cpus.c

```diff
@@ -402,6 +402,21 @@
 // MCPU-SIFIVE-P670-SAME: "-target-feature" "+zvkt"
 // MCPU-SIFIVE-P670-SAME: "-target-abi" "lp64d"
 
+// RUN: %clang -target riscv32 -### -c %s 2>&1 -mcpu=rp2350-hazard3 | FileCheck -check-prefix=MCPU-HAZARD3 %s
+// MCPU-HAZARD3: "-target-cpu" "rp2350-hazard3"
+// MCPU-HAZARD3-SAME: "-target-feature" "+m"
+// MCPU-HAZARD3-SAME: "-target-feature" "+a"
+// MCPU-HAZARD3-SAME: "-target-feature" "+c"
+// MCPU-HAZARD3-SAME: "-target-feature" "+zicsr"
+// MCPU-HAZARD3-SAME: "-target-feature" "+zifencei"
+// MCPU-HAZARD3-SAME: "-target-feature" "+zcb"
+// MCPU-HAZARD3-SAME: "-target-feature" "+zcmp"
+// MCPU-HAZARD3-SAME: "-target-feature" "+zba"
```

Comment:
What's LLVM's opinion on the relation between B and ZbaZbbZbs? Does flagging ZbaZbbZbs without B still allow people to check `#ifdef __riscv_b` in their code?

This tapeout of Hazard3 predates the ratification of B as this combination, so `misa.b` is not set. However `misa.b` is a one-way implication: `misa.b` implies B, but `!misa.b` does not imply !B. Per the ISA manual:

>  When "B" bit is 1, the implementation supports the instructions provided by the Zba, Zbb, and Zbs extensions. When "B" bit is 0, it indicates that the implementation may not support one or more of the Zba, Zbb, or Zbs extensions.

Based on this I would say this version of Hazard3 does implement the B extension, just not `misa.b`. It's a bit of a broader question about LLVM though.

Reply:
LLVM does not currently set __riscv_isa_b when Zba, Zbb, and Zbs are specified by themselves. Retroactively adding new extensions implied by older extension causes things like -fno-integrated-assembler to break if clang knows about B but the GNU assembler on your system doesn't.

Reply:
Understood, thanks. Conversely, does specifying B on its own still allow feature test via the `__riscv_zba` etc macros?  If not, what is the preferred way to detect the presence of e.g. Zba? The context is that I want to avoid mis-detection of the core not having ZbaZbbZbs due to future code only checking the B alias for those extensions, and vice versa.

If the preferred construct is something like `#if defined(__riscv_b) || defined(__riscv_zba)` then the extensions specified in this PR look correct to me. However that sort of code is liable to break next time RISC-V introduces yet another extension alias  (see also A vs Zalrsc/Zaamo, C vs Zca/Zcf/Zcd).

Reply:
> Understood, thanks. Conversely, does specifying B on its own still allow feature test via the __riscv_zba etc macros?

Yes. B does imply Zba, Zbb, and Zbs so those macros will get set.

Reply:
Thanks. I got @lenary to explain the arch string expansion stuff to me offline, and I think I am clear on this now. I'm satisfied that not setting `FeatureStdExtB` is not going to cause any issues with feature detect in the future as long as people don't become too reliant on testing `__riscv_b` only.

Thanks again for the clarification.